File: 0881_boats_to_save_people.swift
// https://leetcode.com/problems/boats-to-save-people/

class Solution {
	func numRescueBoats(_ people: [Int], _ limit: Int) -> Int {
		var people = people.sorted { $0 >= $1 }
		var boats = 0

		while people.count > 0 {
			boats += 1
			let firstPerson = people.removeFirst()
			let rest = limit - firstPerson

			if people.count > 0 {
				if rest >= people.first! {
					people.removeFirst()
				} else if rest < people.last! {
					continue
				} else {
					var l = 0
					var r = people.count - 1
					while l <= r {
						let mid = (l + r) / 2
						let midPerson = people[mid]
						if midPerson == rest {
							l = mid
							r = mid - 1
						} else if midPerson < rest {
							r = mid - 1
						} else {
							l = mid + 1
						}
					}
					people.remove(at: l)
				}
			}
		}

		return boats
	}
}

File: 0881_boats_to_save_people.kt
// https://leetcode.com/problems/boats-to-save-people/description/

class Solution {
    fun numRescueBoats(people: IntArray, limit: Int): Int {
        people.sort()
        var l = 0
        var r = people.size - 1
        var boats = 0
        while (l <= r) {
            if (people[l] + people[r] <= limit) {
                l++
                r--
            } else {
                r--
            }
            boats++
        }
        return boats
    }
}

// Not good solution
class Solution2 {
    fun numRescueBoats(people: IntArray, limit: Int): Int {
        val sortedPeople = arrayListOf<Int>()
        for (person in people) {
            sortedPeople.add(person)
        }
        sortedPeople.sortDescending()
        var boats = 0
        while (sortedPeople.isNotEmpty()) {
            boats++
            val firstPerson = sortedPeople[0]
            sortedPeople.removeAt(0)
            val rest = limit - firstPerson
            var l = 0
            var r = sortedPeople.size - 1
            while (l <= r) {
                val mid = (l + r) / 2
                val midPerson = sortedPeople[mid]
                if (midPerson <= rest) {
                    r = mid - 1
                } else {
                    l = mid + 1
                }
            }
            if (l < sortedPeople.size) {
                sortedPeople.removeAt(l)
            }
        }
        return boats
    }
}
